mpdev/internal/apply/deployment_manager.go

```diff
@@ -0,0 +1,148 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package apply
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+)
+
+// DeploymentManagerAutogenTemplate generates a deployment manager template
+// given an autogen.yaml file.
+type DeploymentManagerAutogenTemplate struct {
+	ResourceShared
+	AutogenFile string
+	PartnerID   string `json:"partnerId"`
+	SolutionID  string `json:"solutionId"`
+
+	outDir string
+}
+
+// Apply generates a deployment manager template from an autogen file.
+func (dm *DeploymentManagerAutogenTemplate) Apply() error {
+	dir, err := ioutil.TempDir("", "autogen")
+	if err != nil {
+		return err
+	}
+	dm.outDir = dir
+
+	fmt.Printf("Generating deployment manager template from autogen\n")
+	err = dm.convertAutogenSchema()
+	if err != nil {
+		return err
+	}
+
+	autogenImg := "gcr.io/cloud-marketplace-tools/dm/autogen"
+	args := []string{"--input_type", "YAML", "--single_input", "/autogen/autogen.yaml",
+		"--output_type", "PACKAGE", "--output", "/autogen"}
+
+	cp := &ContainerProcess{
+		containerImage: autogenImg,
+		processArgs:    args,
+		mounts:         []string{fmt.Sprintf("type=bind,src=%s,dst=/autogen", dm.outDir)},
+	}
+	cmd := cp.getCommand()
+	cmd.Stderr = os.Stderr
+	cmd.Stdout = os.Stdout
+
+	err = cmd.Run()
+	if err != nil {
+		return errors.Wrap(err, "failed to execute autogen")
+	}
+
+	fmt.Printf("Wrote autogen output to directory: %s\n", dm.outDir)
+
+	return nil
+}
+
+func (dm *DeploymentManagerAutogenTemplate) convertAutogenSchema() error {
+	inputFile, err := os.Open(dm.AutogenFile)
+	if err != nil {
+		return err
+	}
+	defer inputFile.Close()
+
+	outFile, err := os.Create(filepath.Join(dm.outDir, "autogen.yaml"))
+	if err != nil {
+		return err
+	}
+	defer outFile.Close()
+
+	// Image name from running `docker build mpdev/autogen -f mpdev/autogen/Dockerfile -t autogen_converter
+	autogenConverterImg := "autogen_converter"
```

Comment:
This will have to be a public image, eventually, right? And we will need to provide a flag to override it if necessary?